lib/sshutils/x11/display.go

```diff
@@ -0,0 +1,196 @@
+	}
+
+	return nil, trace.NewAggregate(tcpErr, unixErr)
+}
+
+// Listen opens an XServer listener
+func (d *Display) Listen() (XServerListener, error) {
+	unixSock, unixErr := d.unixSocket()
+	if unixErr == nil {
+		l, err := net.ListenUnix("unix", unixSock)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return &xserverUnixListener{l}, nil
+	}
+
+	tcpSock, tcpErr := d.tcpSocket()
+	if tcpErr == nil {
+		l, err := net.ListenTCP("tcp", tcpSock)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return &xserverTCPListener{l}, nil
+	}
+
+	return nil, trace.NewAggregate(tcpErr, unixErr)
+}
+
+// xserverUnixSocket returns the display's associated unix socket.
+func (d *Display) unixSocket() (*net.UnixAddr, error) {
+	// For x11 unix domain sockets, the hostname must be "unix" or empty. In these cases
+	// we return the actual unix socket for the display "/tmp/.X11-unix/X<display_number>"
+	if d.HostName == "unix" || d.HostName == "" {
+		sockName := filepath.Join(os.TempDir(), x11SocketDir, fmt.Sprintf("X%d", d.DisplayNumber))
+		return net.ResolveUnixAddr("unix", sockName)
+	}
+	return nil, trace.BadParameter("display is not a unix socket")
+}
+
+// xserverTCPSocket returns the display's associated tcp socket.
+// e.g. "hostname:<6000+display_number>"
+func (d *Display) tcpSocket() (*net.TCPAddr, error) {
+	if d.HostName == "" {
+		return nil, trace.BadParameter("hostname can't be empty for an XServer tcp socket")
+	}
+
+	port := fmt.Sprint(d.DisplayNumber + x11BasePort)
+	rawAddr := net.JoinHostPort(d.HostName, port)
+	addr, err := net.ResolveTCPAddr("tcp", rawAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return addr, nil
+}
+
+// GetXDisplay retrieves and validates the local XServer display
+// set in the environment variable $DISPLAY.
+func GetXDisplay() (Display, error) {
+	displayString := os.Getenv(DisplayEnv)
+	if displayString == "" {
+		return Display{}, trace.BadParameter("$DISPLAY not set")
+	}
+
+	display, err := ParseDisplay(displayString)
+	if err != nil {
+		return Display{}, trace.Wrap(err)
+	}
+	return display, nil
+}
+
+// ParseDisplay parses the given display value and returns the host,
+// display number, and screen number, or a parsing error. display must be
+//in one of the following formats - hostname:d[.s], unix:d[.s], :d[.s], ::d[.s].
+func ParseDisplay(displayString string) (Display, error) {
+	if displayString == "" {
+		return Display{}, trace.BadParameter("display cannot be an empty string")
+	}
+
+	// check the display for illegal characters in case of code injection attempt
+	allowedSpecialChars := ":/.-_" // chars used for hostname or display delimiters.
+	for _, c := range displayString {
+		if !(unicode.IsLetter(c) || unicode.IsNumber(c) || strings.ContainsRune(allowedSpecialChars, c)) {
+			return Display{}, trace.BadParameter("display contains invalid character %q", c)
+		}
+	}
+
+	// Parse hostname.
+	colonIdx := strings.LastIndex(displayString, ":")
```

Comment:
Nothing to change now, but I suspect `strings.Cut()` will be a nice cleanup here when Go 1.18 is out.